### PR TITLE
[SPARK-9963] [ML] RandomForest cleanup: replace predictNodeIndex with predictImpl

### PR DESCRIPTION
**Pull request by lkhamsurenl-zz**
predictNodeIndex is moved to LearningNode and renamed predictImpl for consistency with Node.predictImpl

### CODE REVIEW

**Comment by trevorcai** [comment]
Took a brief look - I have a quick high level thought.
The motivation behind the patch is to replace the redundant functionality of `predictNodeIndex` with the existing `Node.predictImpl` function. Therefore, I'd suggest overloading the `Node.predictImpl` function with the `binnedFeatures` and `splits` arguments and roughly mirroring its current implementation and semantics (just with different arguments provided to `shouldGoLeft`).

That creates consistent semantics of simulating prediction, instead of the split implementations that are currently in the master branch.

**Comment by lkhamsurenl** [comment]
Sorry for the late response! I see what you're saying. The difficulty with that approach I am having is in inside def binSeqOp(agg: Array[DTStatsAggregator], baggedPoint: BaggedPoint[TreePoint]): Array[DTStatsAggregator], where the function is called: treeToNodeToIndexInfo: Map[Int, Map[Int, NodeIndexInfo]] uses nodeIndex to get the NodeInfoIndex, which requires us to know the nodeIndex. If we  convert the LearningNode to Node then implement the predictImpl() there is no nodeIndex I believe. 
Correct me if I'm wrong

**Comment by jkbradley** [comment]
ok to test

File: mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala [hunk 1]
```diff
@@ -279,6 +279,38 @@ private[tree] class LearningNode(
...
+    if (this.isLeaf || this.split.isEmpty) {
+      this
+    } else {
+      val split = this.split.get
+      val featureIndex = split.featureIndex
+      val splitLeft = split.shouldGoLeft(binnedFeatures(featureIndex), splits(featureIndex))
+      if (this.leftChild.isEmpty) {
+        // Not yet split. Return next layer of nodes to train
+        if (splitLeft) {
+          leftChild.get
```
Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala]: This won't work.  I'm running the tests to show the failure.

This makes me realize my original plan (to return the Node itself) won't work; we'll need to stick with returning the ID.  But I like having the code in LearningNode.  Can you please update this method so it returns the ID, not the node?

I'll update the JIRA to match.  Thanks!

Comment (jkbradley) [inline mllib/src/main/scala/org/apache/spark/ml/tree/Node.scala]: Huh...maybe it does work.  Maybe this part of the code is no longer needed.  Still, can you please update as requested to be safe, and we can keep this in mind for potentially simplifying this method later on?

**Comment by jkbradley** [comment]
Thanks for the updates!  LGTM pending tests

Can you please add a little description (1st comment in this PR), at least to say that predictNodeIndex has been moved to inside LearningNode and renamed to predictImpl?  That will become the commit message.

**Comment by jkbradley** [comment]
spurious failure, retesting

**Comment by lkhamsurenl** [comment]
>>> I'll take another look at [#8609] now. Can you please close this duplicate PR?
Done.
>>> Can you please add a little description?
Done.

Please let me know if I need to update anything else.
Thanks

**Comment by jkbradley** [comment]
LGTM
Merging with master.
Thanks for the PR!

**Pull request closed without merging**